File: app/src/main/java/com/ytempest/wanandroid/base/presenter/BasePresenter.java
package com.ytempest.wanandroid.base.presenter;

import com.ytempest.wanandroid.base.view.IView;
import com.ytempest.wanandroid.interactor.impl.BaseInteractor;

import java.lang.reflect.Proxy;

/**
 * @author heqidu
 * @since 2020/6/28
 */
public class BasePresenter<View extends IView> implements IPresenter {

    private View mSrcView;
    protected View mView;
    protected final BaseInteractor mInteractor;

    public BasePresenter(BaseInteractor interactor) {
        mInteractor = interactor;
    }

    @Override
    public <V extends IView> void attachView(V view) {
        mSrcView = (View) view;
        mView = (View) Proxy.newProxyInstance(view.getClass().getClassLoader(), view.getClass().getInterfaces(),
                (proxy, method, args) -> {
                    if (mSrcView != null) {
                        return method.invoke(mSrcView, args);
                    }
                    return null;
                });
    }

    @Override
    public void detach() {
        mSrcView = null;
    }
}

File: app/src/main/java/com/ytempest/wanandroid/base/presenter/BasePresenter.kt
package com.ytempest.wanandroid.base.presenter

import com.ytempest.wanandroid.base.view.IView
import com.ytempest.wanandroid.interactor.impl.BaseInteractor
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy


/**
 * @author heqidu
 * @since 21-2-7
 */
abstract class BasePresenter<View : IView> : IPresenter {


    protected val mInteractor: BaseInteractor
    private var mSrcView: View? = null
    protected lateinit var mView: View

    constructor(interactor: BaseInteractor) {
        mInteractor = interactor
    }

    override fun <V : IView> attachView(view: V) {
        mSrcView = view as View
        mView = mSrcView as View
        // FIXME
        if (true) {
            return
        }
        Proxy.newProxyInstance(view.javaClass.classLoader,
                view.javaClass.interfaces,
                object : InvocationHandler {
                    override fun invoke(proxy: Any?, method: Method?, args: Array<out Any>?): Any {
                        if (mSrcView != null) {
                            return method!!.invoke(proxy, *(args ?: emptyArray()))
                        }
                        return Any()
                    }
                }) as View
    }

    override fun detach() {
        mSrcView = null
    }
}
